File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode head = new ListNode(), ptr = head;

        while (list1 != null && list2 != null) {
            ListNode node = new ListNode();
            if (list1.val < list2.val) {
                node.val = list1.val;
                list1 = list1.next;
            } else {
                node.val = list2.val;
                list2 = list2.next;
            }
            ptr.next = node;
            ptr = ptr.next;

        }

        if (list1 != null) {
            ptr.next = list1;
        }

        if (list2 != null) {
            ptr.next = list2;
        }

        return head.next;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-0.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode) {
    if head == nil || head.Next == nil {
        return
    }

    // locate mid of list
    mid, temp := head, head.Next
    for temp != nil && temp.Next != nil {
        mid = mid.Next
        temp = temp.Next.Next
    }

    // reverse second half of list
    var end *ListNode
    curr := mid.Next
    for curr != nil {
        next := curr.Next
        curr.Next = end
        end = curr
        curr = next
    }
    // disconnect both halves
    mid.Next = nil

    start := head
    // connect first node of first list to first node of second list
    // then move to next nodes of both lists
    for end != nil {
        startNext := start.Next
        endNext := end.Next

        start.Next = end
        end.Next = startNext

        start = startNext
        end = endNext
    }
}
